### PR TITLE
Refactor ipfs get

### PR DESCRIPTION
**Pull request by rht**

### CODE REVIEW

File: unixfs/tar/writer.go [hunk 1]
```diff
@@ -31,12 +32,18 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 	// use a buffered writer to parallelize task
 	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
```
Comment (rht) [inline unixfs/tar/writer.go]: This is not necessary if https://github.com/ipfs/go-ipfs/blob/master/unixfs/tar/writer.go#L123 is replaced with `dagr.WriteTo(w.TarW)`?

Comment (rht) [inline unixfs/tar/writer.go]: Rather, `unixfs/io/dagreader.go` should be rewritten with `bufio` to get Flush automatically.

Comment (whyrusleeping) [inline unixfs/tar/writer.go]: why would the dagreader need flush?

Comment (rht) [inline unixfs/tar/writer.go]: ok it is not necessary for dagreader's `WriteTo` method.

File: unixfs/tar/writer.go [hunk 2]
```diff
@@ -21,79 +21,100 @@ import (
...
+	// compression determines whether to use gzip compression.
+	gzw, err := gzip.NewWriterLevel(pipew, compression)
 	if err != nil {
 		return nil, err
 	}
 
+	w := &Writer{
+		Dag: dag,
+		GzW: gzw,
+		ctx: ctx,
```
Comment (jbenet) [inline unixfs/tar/writer.go]: yeah i debating putting the ctx here too. I opted for the function as it is only needed there-- otherwise the reader can run indefinitely.

File: unixfs/tar/writer.go [hunk 3]
```diff
@@ -21,79 +21,100 @@ import (
...
+// DagArchive is equivalent to `ipfs getdag $hash | maybe_tar | maybe_gzip`
+func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService, useTar bool, compression int) (io.Reader, error) {
 
 	_, filename := path.Split(name)
 
 	// need to connect a writer to a reader
 	piper, pipew := io.Pipe()
 
-	// use a buffered writer to parallelize task
-	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
```
Comment (jbenet) [inline unixfs/tar/writer.go]: the bufio.Writer is here so that we do not need to proceed in lockstep. i think the bufio.Writer will make it so that a worker goroutine can be fetching dag parts and writing them, and a different goroutine can be reading them. the bufio.Writer also has a reasonably big buffer (~3-5 blocks)

File: unixfs/tar/writer.go [hunk 4]
```diff
@@ -21,79 +21,100 @@ import (
...
-		if err != nil {
-			return nil, err
+func (w *Writer) writeDir(nd *mdag.Node, fpath string) error {
+	if w.TarW != nil {
+		if err := writeDirHeader(w.TarW, fpath); err != nil {
+			return err
+		}
+	} else if w.GzW != nil {
+		// gz only case
+		return uio.ErrIsDir
```
Comment (jbenet) [inline unixfs/tar/writer.go]: i think this is getting more confusing than before. the previous function leaved the reader in a single state, and then all ops were straightforward writes. now the conditionals are happening here. Can this go back to using `NewWriter` to set options?

File: unixfs/tar/writer.go [hunk 5]
```diff
@@ -21,79 +21,100 @@ import (
...
+	} else {
+		// no compression / archive case
+		// thirdparty/tar/extractor.go uses 0755 but dir header in this file uses 0777
+		if err := os.MkdirAll(fpath, 0777); err != nil {
+			return err
```
Comment (jbenet) [inline unixfs/tar/writer.go]: I think this is scope creep. the point behind my `DagArchive` reader was to make it work with purely with io streams. now this means a configuration choice might write to the fs. i dont like that-- that should happen separately in another piece.

**Comment by jbenet** [comment]
I appreciate the intent to simplify, and always happy to see code removed, i think some changes here though make the codebase more convoluted (particularly the `DagWriter` changes). can they be done as part of a separate thing? it is trivial and more flexible to stack io.Readers and io.Writers (and often they're zero copy)

File: core/commands/get.go [hunk 6]
```diff
@@ -64,14 +60,34 @@ may also specify the level of compression by specifying '-l=<1-9>'.
...
+					_, outPath = gopath.Split(p.String())
+					outPath = gopath.Clean(outPath)
+				}
+
+				if err := utar.DagFS(ctx, dn, node.DAG, outPath); err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+				return
+			}
```
Comment (jbenet) [inline core/commands/get.go]: :( i moved all this stuff out of this function and into its own because it was very confusing here.

**Comment by jbenet** [comment]
also please tend towards decomposition. larger functions with lots of symbols floating in the same scope are much more error prone as people come in and modify things. smaller/clearer functions are easier to understand and survive better modifications by many people. this PR is putting some stuff back into large functions, after i decomposed it out recently

File: unixfs/tar/writer.go (hunk 2 above)
Comment (rht) [inline unixfs/tar/writer.go]: I tried putting here because it (and dag) doesn't change after initialization.
To compensate, there is `w.Close()` at L56.

File: unixfs/tar/writer.go (hunk 4 above)
Comment (rht) [inline unixfs/tar/writer.go]: Sorry the pr is still in transition state. I was trying to extract out the common node walk pattern (like filepath.walk) between DagArchive, DagZip (previously getZip), and Dag2FS (previously DagArchive, tar.Extractor).

Now I think it is possible to do the code selection at initialization stage (i.e. `NewWriter`). This coming soon.

File: unixfs/tar/writer.go (hunk 5 above)
Comment (rht) [inline unixfs/tar/writer.go]: Sorry will put this back into initialization, just realized of a way to do this without using a "walkFn" with `interface{}`.

File: unixfs/tar/writer.go (hunk 3 above)
Comment (rht) [inline unixfs/tar/writer.go]: Ok, while DagReader's buffer size is as big as a dag node.

File: core/commands/get.go (hunk 6 above)
Comment (rht) [inline core/commands/get.go]: simplified

**Comment by rht** [comment]
I will revert the DagFS back to DagArchive | tar.extract.
The reason is, even if a merkledag node walker exist, it will for ipfs-shell to depend on go-ipfs.
Tar is necessary for transport format.

File: unixfs/tar/writer.go (hunk 2 above)
Comment (rht) [inline unixfs/tar/writer.go]: @jbenet Is `w.Close()` redundant with `pipew.Close()`?

**Comment by rht** [comment]
RFCR

File: unixfs/tar/writer.go [hunk 7]
```diff
@@ -136,16 +152,37 @@ func (w *Writer) WriteNode(ctx cxt.Context, nd *mdag.Node, fpath string) error {
...
+	size := 2 * uint64(tarBlockSize) // tar root padding
+
+	w := &Writer{
+		Dag: dag,
+		ctx: ctx,
+	}
+
+	w.HandleFile = func(dagr *uio.DagReader, fpath string) error {
+		//unixSize := pb.GetFilesize()
+		unixSize := dagr.Size()
```
Comment (rht) [inline unixfs/tar/writer.go]: TODO: this is expensive?

File: unixfs/tar/writer.go [hunk 8]
```diff
@@ -61,91 +64,127 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
...
+	w.HandleFile = func(nd *mdag.Node, pb *upb.Data, fpath string) error {
+		unixSize := pb.GetFilesize()
+		size += tarBlockSize + unixSize + (tarBlockSize - unixSize%tarBlockSize)
+		return nil
+	}
+
+	if err := w.NodeWalk(nd, ""); err != nil {
+		return 0, err
+	}
+	return size, nil
```
Comment (jbenet) [inline unixfs/tar/writer.go]: this is going to write out the whole thing (including pulling out all the objects) just to get the size. is this worth it?

seems like it can also be achieved outside (i.e. not with internal manipulation of the writer) with something like this:

```go
func GetTarSize(...) (uint64, error) {
  cw := CountingWriter(io.Discard) // and discard all the written output.
  w := tar.Writer(..., cw)
  w.WriteDag(root) // do all the writes
  size := cw.Count()
}
```

File: unixfs/tar/writer.go [hunk 9]
```diff
@@ -61,91 +64,127 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
...
+			dagr := uio.NewDataFileReader(w.ctx, nd, pb, w.Dag)
+			_, err := dagr.WriteTo(w.w)
 			return err
 		}
 
-		npath := path.Join(fpath, nd.Links[i].Name)
-		if err := w.WriteNode(ctx, child, npath); err != nil {
+		w.HandleDir = func(fpath string) error {
+			return uio.ErrIsDir
+		}
```
Comment (jbenet) [inline unixfs/tar/writer.go]: if this is not taring output, then it shouldn't be part of the `tar.Writer`. 

and this callback stuff is messy. we should be _composing_ writers, instead of putting both functionalities into one function/struct. why isnt this just two different types? composing writers is cheap, and clear.

```go
// a different func in some other package
func getDagWriter(archive bool, compression int, w io.Writer) io.Writer {
  if compression != gzip.NoCompression {
    w = gzip.Writer(w, compression) // only does the gzip stuff
  }

  if archive {
    return tar.Writer(w, compression) // only does the tar stuff
  }

  return  // what goes here? not archiving nor compressing. 
}
```

File: unixfs/tar/writer.go (hunk 8 above)
Comment (rht) [inline unixfs/tar/writer.go]: The size needs to be computed before the write `w.WriteDag(root)`.

Comment (jbenet) [inline unixfs/tar/writer.go]: is this just for the progress bar? i dont think forcing the sender to look through the entire dag before sending a size is a good idea. it might be _huge_ and take many minutes to compute just the size.

if there's no other way, perhaps the progress bar can estimate and make it clear it's an estimate of the size. as long as it doesn't look borked (i.e. negative numbers, or `bigger/smaller`) and we say "approx X MB" or "est X MB" or even "~X MB" it can be ok.

(relatedly, @krl had suggested calculating a lot of information about dags locally and keeping it around, things like depth, # of nodes, fanout, etc. but i think that's too far off for this. approx result may be better).

Comment (whyrusleeping) [inline unixfs/tar/writer.go]: @rht did we run into trouble just using the 'size' of the dag (as opposed to the true file size) ? I know its not as accurate as we want, but its free information.

Comment (rht) [inline unixfs/tar/writer.go]: @whyrusleeping for a dag with 2000 files, This will be ~ 1MB of inaccuracy. And will be further far off with dag deduplications.

Alternatively is to send dag archive instead of tar. This way the sender doesn't have to undo the deduplication when creating the tar stream.

Comment (whyrusleeping) [inline unixfs/tar/writer.go]: @rht in your opinion, what would be the path of least resistance that also doesnt require multiple DAG traversals?

Comment (rht) [inline unixfs/tar/writer.go]: Excluding the '-C' (compressed) case:
1. use dag size, use tar as transport
    + doesn't require dag traversal
    - is inaccurate, due to mismatch with tar paddings & deduplication
2. use tar size, use tar as transport
    + accurate for all cases
    - require dag traversal
3. compute and include (optional) total unixfs tar size into hash metadata when doing 'ipfs add'

Though I think the best option is
4. use dag size, use dag archive as transport
    + accurate for all cases
    + reduce transport load because dag stays deduplicated
    - format doesn't exist yet

Comment (whyrusleeping) [inline unixfs/tar/writer.go]: could we could use the dag size, and adjust the total size clientside as we receive the data?

Comment (rht) [inline unixfs/tar/writer.go]: The client (ipfs-shell, or the postRun func in ipfs get) sees a tar stream, which means it sees the size of each files, but it doesn't know how each file is associated with a given dag node. So size adjust is not possible.

Comment (jbenet) [inline unixfs/tar/writer.go]: @rht would it help if -- when returning the tar -- we also returned the dag root's cumulative size? we may be able to use this as an estimate, even if it isn't a precise one.

Comment (jbenet) [inline unixfs/tar/writer.go]: alternatively, if you want to start in dag archive, we can do it. i've been meaning to for a while. it would make a lot of things nice.

Comment (rht) [inline unixfs/tar/writer.go]: Basically, use option 1 for now, then switch to option 5 later? What about option 3, how hard is it to add metadata?

Comment (jbenet) [inline unixfs/tar/writer.go]: `1 -> 5` sounds good to me!

I would like to not depend more unixfs. metadata is not trivial. Also the archive / transfer format changing would obviate the work on that.

Comment (rht) [inline unixfs/tar/writer.go]: This would mean the transport format is not swap-able?

How easy is it to change such e.g. both the client transport (currently tar) and archive format (flatfs) are tar?

Comment (jbenet) [inline unixfs/tar/writer.go]: no what i mean is that calculating "tar-ed size" for every single possible root will be obviated when we do use the dag archive as a transport

easy to change transport. repo format not so easy (needs migrations).

Comment (rht) [inline unixfs/tar/writer.go]: For metadata of the tar-ed (and even tar.gz) unixfs size, what about https://github.com/ipfs/go-ipfs/issues/612? And what about using ipfs-ld?

By swapping the transport/repo format, I meant if the other parts of the libs are to be repurposed.
ok, even if it is easy to change/upgrade client transport, the entire network must use only 1 format at a time (the same format for both `ipfs add` and `ipfs get`).

**Comment by rht** [comment]
The refactor has been made much smaller that the PR is safe now.

File: unixfs/tar/writer.go [hunk 10]
```diff
@@ -32,16 +33,30 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
...
 	}
 
 	// write all the nodes recursively
 	go func() {
-		if err := w.WriteNode(ctx, nd, filename); err != nil {
-			pipew.CloseWithError(err)
-			return
+		if !archive && compression != gzip.NoCompression {
+			// the case when the node is a file
+			dagr, err := uio.NewDagReader(w.ctx, nd, w.Dag)
```
Comment (jbenet) [inline unixfs/tar/writer.go]: do we actually know it's a file here? and not a directory?

Comment (rht) [inline unixfs/tar/writer.go]: `uio.NewDagReader` will check and return err if it's not a file.

Comment (rht) [inline unixfs/tar/writer.go]: (I think the naming could have been clarified, but there is also `uio.NewDataFileReader` which doesn't check )

File: unixfs/tar/writer.go [hunk 11]
```diff
@@ -32,16 +33,30 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
...
+			if err != nil {
+				pipew.CloseWithError(err)
+				return
+			}
+
+			if _, err := dagr.WriteTo(w.w); err != nil {
```
Comment (jbenet) [inline unixfs/tar/writer.go]: this is hacky-- the `not-tar` case does not really belong with the tar case. the "node is a file and no archive" case should be handled outside of this. maybe we can re-write this `DagArchive` function as:

```go
// DagArchive is equivalent to `ipfs getdag $hash | maybe_tar | maybe_gzip`
func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService, archive bool, compression int) (io.Reader, error) {

  _, filename := path.Split(name)

  // need to connect a writer to a reader
  piper, pipew := io.Pipe()

  // use a buffered writer to parallelize task
  bufw := bufio.NewWriterSize(pipew, DefaultBufSize)

  // write all the nodes recursively
  go func() {
    if !archive && compression != gzip.NoCompression {
      // the case when the node is a file.
      dagr, err := uio.NewDagReader(w.ctx, nd, w.Dag)
      if err != nil {
        pipew.CloseWithError(err)
        return
      }

      if _, err := dagr.WriteTo(bufw); err != nil {
        pipew.CloseWithError(err)
        return
      }

    } else {
      // construct the tar writer
      tarw, err := NewWriter(ctx, dag, archive, compression, bufw)
      if err != nil {
        return nil, err
      }

      if err := tarw.WriteNode(nd, filename); err != nil {
        pipew.CloseWithError(err)
        return
      }
    }

    if err := bufw.Flush(); err != nil {
      pipew.CloseWithError(err)
      return
    }

    w.Close()
    pipew.Close() // everything seems to be ok.
  }()

  return piper, nil
}
```

and pull it out of the `unixfs/tar` package into something else (maybe `unixfs/archive`?)

Comment (rht) [inline unixfs/tar/writer.go]: "node is a file and no archive" -> then this is not in `unixfs/archive` either.
The 'hack' chunk of the code is meant to wrap `uio.NewDagReader` with a gzip compressor, into another reader.

Comment (rht) [inline unixfs/tar/writer.go]: Currently this gzip wrapper is tangled into tar. Will decompose.

Comment (jbenet) [inline unixfs/tar/writer.go]: yeah gziping is pseudo archive. it could sort of work... i think it's almost there, maybe take the function i gave for now and move on to other things. it's already cleaner

**Comment by jbenet** [comment]
@rht RFM? (if so, just add the label "Ready For Merge" and assign it to me.

**Comment by jbenet** [comment]
@rht: also, maybe can start pushing branches + PRing from within this repo? that way we can collab directly on branches?

**Comment by jbenet** [comment]
> (if so, just add the label "ready to merge" and assign it to me.

(let's try this out as a way to do "hand off" and proper signaling -- cc @whyrusleeping)

**Comment by whyrusleeping** [comment]
> (let's try this out as a way to do "hand off" and proper signaling -- cc @whyrusleeping)

me gusta.
